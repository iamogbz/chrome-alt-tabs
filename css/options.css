body {
    margin: 0;
    padding: 0;
    border: none;
    position: relative;
    font-family: Helvetica, Arial, Serif;
}
a {
    text-decoration:none;
    color:inherit;
}
a:hover {
    color: darkred;
}
header, footer , .container {
    display: block;
    width: calc(100% - 32px);
    padding: 16px;
}
header {
    color: white;
    height: 48px;
    background-color: darkred;
    display: none;
}
header .logo {
    height: 100%;
    display: inline-block;
    float:left;
    border-radius: 4px;
}
header .title {
    margin-left: 32px;
    font-size: 2em;
    line-height: 1.5em;
    float:left;
    display: inline-block;
    height:100%;
}
.container, .content {
    position:relative;
}
.content {
    display:block;
    max-width: 768px;
    margin: 0 auto;
    padding-bottom: 16px;
}
#commands .cmd {
    line-height:32px;
    margin-left: -24px;
}
.sc-cmd {
    color: white;
    background-color: darkred;
    padding: 4px 8px;
    border-radius:4px;
}
.sc-name {
    display: none;
    font-weight: bold;
    margin-left: 8px;
}
.sc-desc {
    margin-left: 8px;
}
footer {
    color: #333;
    background-color: #eee;
    position: absolute;
    bottom:0;
}
#change-btn {
    float: right;
}
button {
    color:white;
    cursor:pointer;
    border: maroon solid 2px;
    border-radius:4px;
    background-color: darkred;
    text-transform:uppercase;
}
button:hover {
    background-color:maroon;
}
button:active {
    color:grey;
}