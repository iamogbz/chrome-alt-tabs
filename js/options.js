var openLink = function () {
    chrome.tabs.update({
        url: this.attributes.href.value
    });
};

document.addEventListener('DOMContentLoaded', function () {
    console.log("DOM loaded");
    // attach listener
    document.getElementById('change-btn').addEventListener("click", openLink);
    // populate commands
    chrome.commands.getAll(function (commands) {
        commands.map(function (cmd) {
            document.getElementById('commands').innerHTML += "<li class='cmd'><span class='sc-cmd'>" + cmd.shortcut + "</span><span class='sc-name'>" + cmd.name + "</span><span class='sc-desc'>" + cmd.description + "</span></li>";
        });
    });
});
