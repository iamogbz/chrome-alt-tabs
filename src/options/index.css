body {
    margin: 0;
    padding: 0;
    border: none;
    position: relative;
    font-family: Helvetica, Arial, Serif;
}

a {
    text-decoration: none;
    color: inherit;
}

a:hover {
    color: darkred;
}

footer,
.container {
    display: block;
    width: calc(100% - 32px);
    padding: 16px;
}

.container,
.content {
    position: relative;
}

.content {
    display: block;
    max-width: 768px;
    min-width: 360px;
    margin: 0 auto;
    padding-bottom: 16px;
}

#commands .cmd {
    line-height: 32px;
    margin-left: -24px;
}

.sc-cmd {
    color: white;
    background-color: darkred;
    padding: 4px 8px;
    border-radius: 4px;
}

.sc-name {
    display: none;
    font-weight: bold;
    margin-left: 8px;
}

.sc-desc {
    margin-left: 8px;
}

footer {
    color: #333;
    background-color: #eee;
    position: relative;
    bottom: 0;
}

#change-btn {
    float: right;
}

button {
    color: white;
    cursor: pointer;
    border: maroon solid 2px;
    border-radius: 4px;
    background-color: darkred;
    text-transform: uppercase;
}

button:hover {
    background-color: maroon;
}

button:active {
    color: grey;
}
