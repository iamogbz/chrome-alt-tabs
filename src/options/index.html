<html>

<head>
    <title>Alt Tabs Options</title>
    {# <link rel="stylesheet" type="text/css" href="css/options.css" /> #}
</head>

<body>
    <header>
        <img class="logo" src="icon.128.png" alt="Alt Tabs">
        <span class="title">Alt Tabs Options</span>
    </header>
    <div id="main-content" class="container">
        <div class="content">
            <a id='change-btn' href="chrome://extensions/configureCommands">
                <button>Change</button>
            </a>
            <h1>Shortcuts</h1>
            <ul id="commands"></ul>
        </div>
    </div>
    <footer>
        <span>Courtesy
            <a href="https://ogbizi.com/labs" target="_blank">Ogbizi Labs</a>
        </span>
    </footer>
    <script src="options.js"></script>
</body>

</html>
