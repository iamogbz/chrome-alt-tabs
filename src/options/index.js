const openLink = () => {
    chrome.tabs.update({
        url: this.attributes.href.value,
    });
};

document.addEventListener("DOMContentLoaded", () => {
    document.getElementById("change-btn").addEventListener("click", openLink);
    chrome.commands.getAll(commands => {
        document.getElementById("commands").innerHTML = commands
            .map(
                cmd =>
                    `<li class='cmd'><span class='sc-cmd'>${
                        cmd.shortcut
                    }</span><span class='sc-name'>${
                        cmd.name
                    }</span><span class='sc-desc'>${
                        cmd.description
                    }</span></li>`,
            )
            .join("");
    });
});
