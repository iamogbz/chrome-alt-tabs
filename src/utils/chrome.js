import { wrapIndex } from "./base";

/**
 * Bind single oncommand listener and switch based on received
 */
const listeners = {};
chrome.commands.onCommand.addListener(command => {
    (listeners[command] || []).forEach(callback => callback());
});
/**
 * Register callback to be run on specific command
 * @param {String} command activation trigger
 * @param {Function} callback to be run
 */
export const onCommand = (command, callback) => {
    if (!Object.keys(listeners).includes(command)) {
        listeners[command] = new Set();
    }
    listeners[command].add(callback);
};

/**
 * Get all registered chrome commands
 */
export const getAllCommands = () =>
    new Promise(resolve => chrome.commands.getAll(resolve));

/**
 * Change current tab location
 */
export const changeTabUrl = url =>
    new Promise(resolve => chrome.tabs.update({ url }, resolve));

/**
 * Bring window to forefront
 * @param {Number} windowId
 */
export const focusOnWindow = windowId =>
    new Promise(resolve => {
        chrome.windows.update(windowId, { focused: true }, resolve);
    });

/**
 * Focus on tab using the unique chrome id
 * @param {Number} tabId unique across multiple windows
 */
export const focusOnTab = tabId =>
    new Promise(resolve => {
        chrome.tabs.update(tabId, { active: true }, resolve);
    });

/**
 * Move tab into a window
 * @param {Number} tabId
 * @param {Number} windowId
 * @param {Number} index position in window, defaults to last
 * @returns {Promise<[Tab]>}
 */
const moveTabToWindow = (tabId, windowId, index = -1) =>
    new Promise(resolve => {
        chrome.tabs.move(tabId, { windowId, index }, resolve);
    });

/**
 * Move multiple tabs into a window restoring position if possible
 * @param {[Tab]} tabs
 * @param {Number} windowId
 * @returns {Promise<[Tab]>}
 */
export const moveTabsToWindow = (tabs, windowId) =>
    Promise.all(
        tabs.map(tab => {
            const index = tab.windowId === windowId ? tab.index : -1;
            return moveTabToWindow(tab.id, windowId, index);
        }),
    );

/**
 * Highlight supplied tab
 * @param {Number} tabId
 * @returns {Promise}
 */
const selectTab = tabId =>
    new Promise(resolve => {
        chrome.tabs.update(tabId, { highlighted: true }, resolve);
    });

/**
 * Select tabs by highlighting
 * @param {[Number]} tabIds list of tabs to select
 * @returns {Promise}
 */
export const selectTabs = tabIds => Promise.all(tabIds.map(selectTab));

/**
 * Get window last focused on
 * @returns {Promise<Window>}
 */
export const getLastFocusedWindow = () =>
    new Promise(resolve => chrome.windows.getLastFocused({}, resolve));

/**
 * Get selected tabs in specified window
 * @param {Number} windowId
 * @returns {Promise<[Tabs]>}
 */
export const getSelectedTabsInWindow = windowId =>
    new Promise(resolve =>
        chrome.tabs.query(
            {
                highlighted: true,
                windowId,
            },
            resolve,
        ),
    );

/**
 * Create window from existing tab
 * @param {Number} tabId tab to move into new window
 * @returns {Promise<Window>}
 */
export const createWindow = tabId =>
    new Promise(resolve => chrome.windows.create({ tabId }, resolve));

/**
 * Get all windows of matching type
 * @param  {...string} windowTypes default to ["normal"]
 * @returns {Promise<[Window]>}
 */
const getAllWindows = (...windowTypes) => {
    if (windowTypes.length === 0) {
        windowTypes.push("normal");
    }
    return new Promise(resolve => {
        chrome.windows.getAll({ windowTypes }, resolve);
    });
};

/**
 * Compare two window to determine order
 * @param {Window} windowA
 * @param {Window} windowB
 */
const compareWindowPositions = (windowA, windowB) => {
    const props = ["left", "top", "width", "height", "id"];
    while (props.length) {
        const prop = props.pop();
        const valueA = windowA[prop];
        const valueB = windowB[prop];
        if (valueA !== undefined && valueB !== undefined && valueA !== valueB) {
            return valueA - valueB;
        }
    }
    return 0;
};

/**
 * Get all window ids sorted
 * @returns {Promise<[number]>}
 */
const getSortedWindowIds = async () => {
    const windows = await getAllWindows();
    windows.sort(compareWindowPositions);
    return windows.map(({ id }) => id);
};

/**
 * Get window in the direction of the position offset from supplied window id
 * @param {Number} windowId
 * @param {Number} offset
 * @returns {Promise<{ id: Number }>}
 */
const getWindowInPositionFrom = async (windowId, offset) => {
    const windowIds = await getSortedWindowIds();
    const index = windowIds.indexOf(windowId) + offset;
    return { id: windowIds[wrapIndex({ index, size: windowIds.length })] };
};

/**
 * Get window after supplied window id
 * @param {Number} id
 * @returns {Promise<{ id: Number }>}
 */
export const getWindowAfter = id => getWindowInPositionFrom(id, 1);

/**
 * Get window before supplied window id
 * @param {Number} id
 * @returns {Promise<{ id: Number }>}
 */
export const getWindowBefore = id => getWindowInPositionFrom(id, -1);
