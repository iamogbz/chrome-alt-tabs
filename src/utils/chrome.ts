import { Tab, Window } from "../types";
import { wrapIndex } from "./base";

/**
 * Bind single oncommand listener and switch based on received
 */
const listeners: { [key: string]: Set<Function> } = {};
chrome.commands.onCommand.addListener(command => {
    (listeners[command] || []).forEach((callback: Function) => callback());
});

/**
 * Register callback to be run on specific command
 */
export const onCommand = (command: string, callback: Function): void => {
    if (!Object.keys(listeners).includes(command)) {
        listeners[command] = new Set();
    }
    listeners[command].add(callback);
};

/**
 * Get all registered chrome commands
 */
export const getAllCommands = (): Promise<chrome.commands.Command[]> =>
    new Promise(resolve => chrome.commands.getAll(resolve));

/**
 * Change current tab location
 */
export const changeTabUrl = (url: string): Promise<Tab> =>
    new Promise(resolve => chrome.tabs.update({ url }, resolve));

/**
 * Bring window to forefront
 */
export const focusOnWindow = (windowId: number) =>
    new Promise(resolve => {
        chrome.windows.update(windowId, { focused: true }, resolve);
    });

/**
 * Focus on tab using unique id; will highlight the tab by default
 */
export const focusOnTab = (tabId: number): Promise<Tab> =>
    new Promise(resolve => {
        chrome.tabs.update(tabId, { active: true }, resolve);
    });

/**
 * Move tab into a window
 */
const moveTabToWindow = (
    tabId: number,
    windowId: number,
    index: number = -1,
): Promise<Tab> =>
    new Promise(resolve => {
        chrome.tabs.move(tabId, { windowId, index }, resolve);
    });

/**
 * Move multiple tabs into a window restoring position if possible
 */
export const moveTabsToWindow = (
    tabs: Tab[],
    windowId: number,
): Promise<Tab[]> =>
    Promise.all(
        tabs.map(tab => {
            const index = tab.windowId === windowId ? tab.index : -1;
            return moveTabToWindow(tab.id, windowId, index);
        }),
    );

/**
 * Highlight supplied tab
 */
const selectTab = (tabId: number): Promise<Tab> =>
    new Promise(resolve => {
        chrome.tabs.update(tabId, { highlighted: true }, resolve);
    });

/**
 * Select tabs by highlighting
 */
export const selectTabs = (tabIds: number[]): Promise<Tab[]> =>
    Promise.all(tabIds.map(selectTab));

/**
 * Get window last focused on
 */
export const getLastFocusedWindow = (): Promise<Window> =>
    new Promise(resolve => chrome.windows.getLastFocused({}, resolve));

/**
 * Get selected tabs in specified window
 */
export const getSelectedTabsInWindow = (windowId: number): Promise<Tab[]> =>
    new Promise(resolve =>
        chrome.tabs.query(
            {
                highlighted: true,
                windowId,
            },
            resolve,
        ),
    );

/**
 * Create window from existing tab
 */
export const createWindow = (tabId: number): Promise<Window> =>
    new Promise(resolve => chrome.windows.create({ tabId }, resolve));

/**
 * Get all windows of matching type, default to ["normal"]
 */
const getAllWindows = (...windowTypes: string[]): Promise<Window[]> => {
    if (windowTypes.length === 0) {
        windowTypes.push("normal");
    }
    return new Promise(resolve => {
        chrome.windows.getAll({ windowTypes }, resolve);
    });
};

/**
 * Compare two window to determine order
 */
const compareWindowPositions = (windowA: Window, windowB: Window) => {
    const props = ["left", "top", "width", "height", "id"];
    while (props.length) {
        const prop = props.pop();
        const valueA = windowA[prop];
        const valueB = windowB[prop];
        if (valueA !== undefined && valueB !== undefined && valueA !== valueB) {
            return valueA - valueB;
        }
    }
    return 0;
};

/**
 * Get all window ids sorted
 */
const getSortedWindowIds = async (): Promise<number[]> => {
    const windows = await getAllWindows();
    windows.sort(compareWindowPositions);
    return windows.map(({ id }) => id);
};

/**
 * Get window in the direction of the position offset from supplied window id
 */
const getWindowInPositionFrom = async (
    windowId: number,
    offset: number,
): Promise<{ id: number }> => {
    const windowIds = await getSortedWindowIds();
    const index = windowIds.indexOf(windowId) + offset;
    return { id: windowIds[wrapIndex({ index, size: windowIds.length })] };
};

/**
 * Get window after supplied window id
 */
export const getWindowAfter = (id: number): Promise<{ id: number }> =>
    getWindowInPositionFrom(id, 1);

/**
 * Get window before supplied window id
 */
export const getWindowBefore = (id: number): Promise<{ id: number }> =>
    getWindowInPositionFrom(id, -1);
