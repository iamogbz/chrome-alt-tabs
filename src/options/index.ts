import { changeTabUrl, getAllCommands } from "../utils/chrome";

document.addEventListener("DOMContentLoaded", async () => {
    const changeShortcutsButton = document.getElementById("change-btn");
    changeShortcutsButton.addEventListener("click", () => {
        const targetAttrs = changeShortcutsButton.attributes;
        changeTabUrl(targetAttrs.getNamedItem("href").value);
    });
    const commands = await getAllCommands();
    document.getElementById("commands").innerHTML = commands
        .map(
            cmd =>
                `<li class='cmd'>
                <span class='sc-cmd'>${cmd.shortcut}</span>
                <span class='sc-name'>${cmd.name}</span>
                <span class='sc-desc'>${cmd.description}</span>
                </li>`,
        )
        .join("");
});
