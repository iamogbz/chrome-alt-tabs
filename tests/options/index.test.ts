import "options";
import * as optionsHTML from "options/index.html";
import * as chromeUtils from "utils/chrome";

const changeTabUrlSpy = jest.spyOn(chromeUtils, "changeTabUrl");
jest.spyOn(chromeUtils, "getAllCommands").mockResolvedValue([
  { shortcut: "ctrl+1", name: "Command1", description: "mock command 1" },
  { shortcut: "ctrl+2", name: "Command2", description: "mock command 2" },
  { shortcut: "ctrl+3", name: "Command3", description: "mock command 3" },
]);

describe("options", () => {
  beforeAll(() => {
    document.open("text/html");
    document.write(optionsHTML);
    document.close();
  });

  it("renders all commands", async () => {
    document.dispatchEvent(new Event("DOMContentLoaded"));
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(document.getElementById("commands").innerHTML).toMatchSnapshot();
  });

  it("changes tab url", async () => {
    document.getElementById("change-btn").click();
    expect(changeTabUrlSpy).toHaveBeenCalledWith(
      "chrome://extensions/configureCommands"
    );
  });
});
