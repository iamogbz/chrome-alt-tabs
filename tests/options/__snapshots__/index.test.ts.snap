// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options renders all commands 1`] = `
"<li class=\\"cmd\\">
                <span class=\\"sc-cmd\\">ctrl+1</span>
                <span class=\\"sc-name\\">Command1</span>
                <span class=\\"sc-desc\\">mock command 1</span>
                </li><li class=\\"cmd\\">
                <span class=\\"sc-cmd\\">ctrl+2</span>
                <span class=\\"sc-name\\">Command2</span>
                <span class=\\"sc-desc\\">mock command 2</span>
                </li><li class=\\"cmd\\">
                <span class=\\"sc-cmd\\">ctrl+3</span>
                <span class=\\"sc-name\\">Command3</span>
                <span class=\\"sc-desc\\">mock command 3</span>
                </li>"
`;
